---
title: "Estimating SNP-heritabili and genetic correlation using LDSC and HDL"
author: "Thi Thuy Dung Nguyen"
date: "30/11/2020"
output: html_document
editor_options: 
  chunk_output_type: console
---
This document is a supplementary file for article:

Genetic heterogeneity and subtypes of major depression

Thuy-Dung Nguyen, M.Sc.(1,2); Arvid Harder, B.Sc.1; Ying Xiong, M.Sc.(1); Kaarina Kowalec, Ph.D.(1,3); Sara Hägg, Ph.D.(1); Na Cai, Ph.D.(4), Ralf Kuja-Halkola, Ph.D.(1); Christina Dalman, M.D.(2); Patrick F Sullivan, M.D. FRANZCP(1,5); Yi Lu, Ph.D.(1,2)

1.	Department of Medical Epidemiology and Biostatistics, Karolinska Institutet, Stockholm, Sweden
2.	Department of Global Public Health, Karolinska Institute, Stockholm, Sweden
3.	College of Pharmacy, University of Manitoba, Winnipeg, Canada
4.	Helmholtz Pioneer Campus, Helmholtz Zentrum München, Neuherberg, Germany
5.	Department of Genetics and Psychiatry, University of North Carolina, Chapel Hill, NC, USA

Corresponding author: 
Yi Lu, Ph.D.
Nobels Väg 12A, 
Solna, Sweden 17177  
lu.yi@ki.se 
Responsible for this code file
Thuy-Dung Nguyen
Nobels Väg 12A, 
Solna, Sweden 17177  
thuy.dung.nguyen@ki.se

In this document, we present a pipeline to 
1) Constructing genetic relationship matrix using GCTA program
2) Case-control GWAS using fastGWA model of GCTA program
3) Esitmate SNP-heritability using linkage disequilibrium score regression (LDSC) 
4) Estimating genetic correlation using High-Definition Likelihood (HDL)

# I. Constructing genetic relationship matrix

## Pruning UKB data, and extracting HAPMAP3 SNPs
In this step, we extract LD-pruned Hapmap3 variants from the full genotype data
HAPMAP contains Hapmap3 variants ID
OUTPUT is to store 22 LD-pruned variant list file for 22 chromosomes. The 22 chromosomes will be joined in the next step

```{eval=FALSE}
HAPMAP="/<path to hapmap3 file>/hapmap3_full.txt"
OUTPUT="/<path/folder to store output files>"

for i in $(seq 1 22); do
echo '#!/bin/bash
#SBATCH --job-name=ldpruning_chr'"${i}"'
#SBATCH --account=<project_name>
#SBATCH --time=00-10:00:00
#SBATCH --partition=core
#SBATCH --ntasks=2'"

module load bioinfo-tools
module load plink2

plink2 --bfile /<path to binary file folder>/ukb_imp_chr${i}_v3_maf_0.01 \
--indep-pairwise 1000 100 0.9 \
--threads 2 \
--extract $HAPMAP \
--out $OUTPUT${i}
" > ldprune.sh
done
```

# Joining the output from each chromosome.
All variant lists from 22 chromosome are joined in 1 file in this step
```{r}
R
library(tidyverse)

final <- tibble()
names <- c()
for(i in 2:22){
  names[i] <- paste("ld_pruned_chr", as.character(i), ".prune.in", sep = "")
  data <- read_tsv(names[i], col_names = FALSE)
  if(i == 1){
      final <- data
  } else {
      final <- bind_rows(final, data)
  }
  
}
write_tsv(final, "/<path>/whole_genome_ldpruned.txt")
```

## Generating the GRM based on the ldpruned set
For UKB, we divide into 300 parts, to reduce memory usage per job to less than 7GB, so each job can run on one core.

LDSET contains the whole genome ld-pruned variants list in the previous step
MBFILE contains list of the bgen files for 22 chromosomes

```{bash}
LDSET="/<path>/whole_genome_ldpruned.txt"
MBFILE="/<path>/chromosomes.txt"
for i in $(seq 1 300); do
echo '#!/bin/bash
#SBATCH --job-name=grm_entire_genome_part'"${i}"'
#SBATCH --account=sens2017519
#SBATCH --time=00-24:00:00
#SBATCH --partition=core
#SBATCH --ntasks=1'"

/<path to gcta directory>/gcta64 \
--mbfile $MBFILE \
--maf 0.01 \
--extract $LDSET \
--make-grm-part 300 ${i} \
--sparse-cutoff 0.05 \
--threads 1 \
--out /<path to folder>/entire_genome_part${i}
" > script.sh
sbatch script.sh
sleep 0.1s
done
```

# 2. GWAS
GWAS was done for all phenotypes using fastGWA moduls in GCTA. Details about the methods in program can be found at:
Jiang et al. (2019) A resource-efficient tool for mixed model association analysis of large-scale data. Nature Genet. 51, 1749–1755 (2019) doi:10.1038/s41588-019-0530-8.
Computation was done on UPPMAX server

GCTA contains GCTA program
GENO contains list of bgen files
CATCOVAR contains categorical covariate vectors
CONTCOVAR contains continuous covariate vectors
SAMPLE contains the sample that corresponse to bgen files
GRM is the folder of GRM in the previous step
PHENOFOLDER contains phenotype data 
GWASFOLDER is to store GWAS result files

## 
```{bash}
GCTA="/<path to gcta directory>/gcta64"
GENO="/<path to bgen list .txt file>/ukb_bgen_list.txt"
CATCOVAR="/<path to categorial covariate file>/sex_fastGWA.txt"
CONTCOVAR="/<path to continuous covariate file>/pc10_age_fastGWA.txt"
SAMPLE="/<path to the .sample file>/ukb2222_imp_v3_addsex.sample"
GRM="/<path to folder containing GRM>/"
PHENOFOLDER="<path/phenotype foler name>"
GWASFOLDER="<path/GWAS output foler name>"

for line in $(ls <path/phenotype foler name> | tee ); do
echo '#!/bin/bash
#SBATCH --job-name='"$line"'
#SBATCH --account=<your project name>
#SBATCH --time=02-00:00:00
#SBATCH --partition=core
#SBATCH --ntasks=16
#SBATCH -C mem256GB'"
$GCTA \
--mbgen $GENO \
--sample $SAMPLE \
--grm-sparse $GRM \
--fastGWA-mlm \
--qcovar $CONTCOVAR \
--covar $CATCOVAR \
--pheno $PHENOFOLDER/$line \
--maf 0.01 \
--threads 16 \
--out $GWASFOLDER/$line
" > $line.sh
chmod 700 $line.sh 
sbatch $line.sh 
done
```

# 3. ESTIMATING SNP-HERITABILITY USING LDSC
Details about the methods and instruction to install the software can be found here:
Bulik-Sullivan BK, Loh P-R, Finucane HK, Ripke S, Yang J, Patterson N, et al. LD Score regression distinguishes confounding from polygenicity in genome-wide association studies. Nature Genetics. 2015;47(3):291-5.

## 3.1 Prepare sumstats for LDSC
The input for LDSC is sumary statistics of SNPs from GWAS. sumstats function in LDSC wrangle the sumstats data in the right format for LDSC to run.
In this code chunk, we loop through all GWAS files in the GWASFOLDER to transform all files into the right format. 

LDSCDIR contains LDSC program
GWASFOLDER contains all GWAS files in the previous step
ALLELE is a file that contains list of SNPs to be matched
SUMSTATSFOLDER is to contain all formated sumstats output files

```{bash}
LDSCDIR="/<path/name of directory with ldsc program>/"
GWASFOLDER="/<path/folder that contains all GWAS files>/"
ALLELE="/<path>/w_hm3.noMHC.snplist"
SUMSTATSFOLDER="/<path/formated sumstats foler name>/"

for line in $(ls <path/folder that contains all GWAS file> | tee ); do
$LDSCDIR/munge_sumstats.py \
--sumstats $GWASFOLDER/$line \
--snp SNP \
--signed-sumstats BETA,0 \
--p P \
--merge-alleles $ALLELE \
--info INFO \
--chunksize 500000 \
--out $SUMSTATFOLDER/$line
done
```

## 3.2 Estimating SNP-heritability 
In out for estimating SNP-heritabily is summary statistics from GWAS in a proper format as done in the sumstats step above. In this code chunk, we looped through all sumstats file and estimated SNP-heritability for all phenotypes.

LDSCDIR contains LDSC program
SUMSTATSFOLDER contains all sumstats files for all phenotypes
REFLD contains 22 files of LD score from a reference population for 22 chromosomes 
OUTFOLDER is to store results files

```{bash}
LDSCDIR="/<path/name of directory with ldsc program>/"
SUMSTATFOLDER="/<path/formated sumstats foler name>/"
REFLD="/<path/to reference LD score folder>/"
OUTFOLDER="/<path/output folder name>/"

for line in $(ls $SUMSTATFOLDER/ | tr '\n' '\n' | grep .gz| tee); do
$LDSCDIR/ldsc.py \
--h2 $SUMSTATFOLDER/$line \
--ref-ld-chr $REFLD \
--w-ld-chr $REFLD \
--out $OUTFOLDER/$line
done
```

# 4. ESTIMATING GENETIC CORRELATION USING HDL
Details about the methods and instruction to install the software can be found here:
Ning, Z., Pawitan, Y. & Shen, X. High-definition likelihood inference of genetic correlations across human complex traits. Nat Genet (2020).
We included the R script used in this analyses in this github project folder but users are recommended to see instruction how to install the program here:
https://github.com/zhenin/HDL/wiki/Installation-and-update

## 4.1 Munge data

```{bash}
HDL_WRANGLE="/<path to HDL program directory>/HDL.data.wrangling.R"
GWASFOLDER="/<path to GWAS folder>/GWAS_subtype_20nov"
eigenarray="/<path/LD reference panel eigenvalues folder>/"

WRANFLESUMSTATFOLDER="/<path/output folder to store formated sumstats files>/"
LOGOUT="/<path/output folder to store formated log files>/"

for line in $(ls $GWASFOLDER/ | tr '\n' '\n' | tee); do
Rscript $HDL_WRANGLE \
gwas.file=$GWASFOLDER/$line \
LD.path=$eigenarray \
SNP=SNP A1=A1 A2=A2 N=N b=BETA se=SE \
output.file=$WRANFLESUMSTATFOLDER/$line \
log.file=$LOGOUT/$line
done
```

## 4.2 Estimating genetic correlation
This code chunk lists all files in the folder WRANFLESUMSTATFOLDER

```{bash}
WRANFLESUMSTATFOLDER="/<path/output folder to store formated sumstats files>/"
ls $WRANFLESUMSTATFOLDER/ | tr '\n' '\n' | grep .rds| tee
```
Folder WRANFLESUMSTATFOLDER contains formated GWAS summary statistics results for 6 phenotypes.
After listing all files in the folder, they are listed in the list mungefile in the order that genetic correlation will be calculated between pheno0 and pheno1, pheno2 and pheno3, and so on.

HDL_RG contain HDL program
eigenarray contains array reference panel
OUTFOLDER is to store HDL result files

```{bash}
HDL_RG="/<path to HDL program directory>/HDL.parallel.run.R"
eigenarray="/<path to eigen array folder>/UKB_imputed_SVD_eigen99_extraction"
WRANFLESUMSTATFOLDER="/<path/output folder to store formated sumstats files>/"
OUTFOLDER="/<path/folder to store HDL results>/"
mungefile=(
pheno0.rds 
pheno1.rds 
pheno2.rds
pheno3.rds
pheno4.rds
pheno5.rds
pheno6.rds
)
order=(0 2 4 6)

for i in ${order[*]}; do
echo '#!/bin/bash
#SBATCH --job-name='"$i"'
#SBATCH --time=02-00:00:00
#SBATCH --partition=core
#SBATCH --ntasks=8
#SBATCH --mem=8G'"

Rscript $HDL_RG \
gwas1.df=$WRANFLESUMSTATFOLDER/${mungefile[$i]} \
gwas2.df=$WRANFLESUMSTATFOLDER/${mungefile[$i+1]} \
LD.path=$eigenarray \
output.file=$OUTFOLDER/${mungefile[$i]} \
numCores=8

" > ${mungefile[$i]}.sh
chmod 700 ${mungefile[$i]}.sh
sbatch ${mungefile[$i]}.sh
rm ${mungefile[$i]}.sh
done
```
